diff --git gcc-9-20190324/gcc/ada/Makefile.rtl gcc-9-20190324/gcc/ada/Makefile.rtl
index 775ab98..636d6b5 100644
--- gcc-9-20190324/gcc/ada/Makefile.rtl
+++ gcc-9-20190324/gcc/ada/Makefile.rtl
@@ -1521,7 +1521,6 @@ ifeq ($(strip $(filter-out %86 linux%,$(target_cpu) $(target_os))),)
   s-tpopsp.adb<libgnarl/s-tpopsp__tls.adb \
   $(TRASYM_DWARF_UNIX_PAIRS) \
   g-sercom.adb<libgnat/g-sercom__linux.adb \
-  s-tsmona.adb<libgnat/s-tsmona__linux.adb \
   a-exetim.adb<libgnarl/a-exetim__posix.adb \
   a-exetim.ads<libgnarl/a-exetim__default.ads \
   s-linux.ads<libgnarl/s-linux.ads \
@@ -2073,7 +2072,6 @@ ifeq ($(strip $(filter-out powerpc% linux%,$(target_cpu) $(target_os))),)
   s-tpopsp.adb<libgnarl/s-tpopsp__tls.adb \
   g-sercom.adb<libgnat/g-sercom__linux.adb \
   $(TRASYM_DWARF_UNIX_PAIRS) \
-  s-tsmona.adb<libgnat/s-tsmona__linux.adb \
   $(ATOMICS_TARGET_PAIRS) \
   $(ATOMICS_BUILTINS_TARGET_PAIRS) \
   system.ads<libgnat/system-linux-ppc.ads
@@ -2297,7 +2295,6 @@ ifeq ($(strip $(filter-out %ia64 linux%,$(target_cpu) $(target_os))),)
   s-taspri.ads<libgnarl/s-taspri__posix-noaltstack.ads \
   g-sercom.adb<libgnat/g-sercom__linux.adb \
   $(TRASYM_DWARF_UNIX_PAIRS) \
-  s-tsmona.adb<libgnat/s-tsmona__linux.adb \
   $(ATOMICS_TARGET_PAIRS) \
   $(ATOMICS_BUILTINS_TARGET_PAIRS) \
   system.ads<libgnat/system-linux-ia64.ads
@@ -2394,7 +2391,6 @@ ifeq ($(strip $(filter-out %x86_64 linux%,$(target_cpu) $(target_os))),)
   s-taspri.ads<libgnarl/s-taspri__posix.ads \
   g-sercom.adb<libgnat/g-sercom__linux.adb \
   $(TRASYM_DWARF_UNIX_PAIRS) \
-  s-tsmona.adb<libgnat/s-tsmona__linux.adb \
   $(ATOMICS_TARGET_PAIRS) \
   $(X86_64_TARGET_PAIRS) \
   system.ads<libgnat/system-linux-x86.ads
diff --git gcc-9-20190324/gcc/ada/adadecode.c gcc-9-20190324/gcc/ada/adadecode.c
index a574f3c..af56dac 100644
--- gcc-9-20190324/gcc/ada/adadecode.c
+++ gcc-9-20190324/gcc/ada/adadecode.c
@@ -29,6 +29,7 @@
  *                                                                          *
  ****************************************************************************/
 
+#include "adaint.h"  /* for a macro version of xstrdup.  */
 
 #if defined(IN_RTS)
 #include "tconfig.h"
@@ -42,8 +43,6 @@
 #include <stdio.h>
 #include <ctype.h>
 
-#include "adaint.h"  /* for a macro version of xstrdup.  */
-
 #ifndef ISDIGIT
 #define ISDIGIT(c) isdigit(c)
 #endif
diff --git gcc-9-20190324/gcc/ada/adaint.c gcc-9-20190324/gcc/ada/adaint.c
index 4a75b59..b9b23e6 100644
--- gcc-9-20190324/gcc/ada/adaint.c
+++ gcc-9-20190324/gcc/ada/adaint.c
@@ -79,6 +79,11 @@
 #include <sys/pstat.h>
 #endif
 
+#if defined (__linux__)
+#define _GNU_SOURCE
+#include <sched.h>
+#endif
+
 #ifdef __PikeOS__
 #define __BSD_VISIBLE 1
 #endif
@@ -91,6 +91,8 @@
 #include <sys/syspage.h>
 #endif
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -208,8 +210,6 @@ UINT __gnat_current_ccs_encoding;
 #include <utime.h>
 #endif
 
-#include "adaint.h"
-
 /* Define symbols O_BINARY and O_TEXT as harmless zeroes if they are not
    defined in the current system. On DOS-like systems these flags control
    whether the file is opened/created in text-translation mode (CR/LF in
@@ -3271,7 +3275,6 @@ __gnat_lwp_self (void)
 #endif
 
 #if defined (__linux__)
-#include <sched.h>
 
 /* glibc versions earlier than 2.7 do not define the routines to handle
    dynamically allocated CPU sets. For these targets, we use the static
diff --git gcc-9-20190324/gcc/ada/adaint.h gcc-9-20190324/gcc/ada/adaint.h
index 2ae8766..0422af3 100644
--- gcc-9-20190324/gcc/ada/adaint.h
+++ gcc-9-20190324/gcc/ada/adaint.h
@@ -316,6 +316,7 @@ extern void   *__gnat_lwp_self			   (void);
 
 /* Routines for interface to required CPU set primitives */
 
+#define _GNU_SOURCE
 #include <sched.h>
 
 extern cpu_set_t *__gnat_cpu_alloc                 (size_t);
diff --git gcc-9-20190324/gcc/ada/cio.c gcc-9-20190324/gcc/ada/cio.c
index dd91a3e..72382a3 100644
--- gcc-9-20190324/gcc/ada/cio.c
+++ gcc-9-20190324/gcc/ada/cio.c
@@ -29,6 +29,8 @@
  *                                                                          *
  ****************************************************************************/
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -38,8 +40,6 @@
 #include "system.h"
 #endif
 
-#include "adaint.h"
-
 /* We need L_tmpnam definition */
 #include <stdio.h>
 
diff --git gcc-9-20190324/gcc/ada/argv.c gcc-9-20190324/gcc/ada/argv.c
index 3249c32..3bbe17a 100644
--- gcc-9-20190324/gcc/ada/argv.c
+++ gcc-9-20190324/gcc/ada/argv.c
@@ -42,6 +42,8 @@
    main program, and these routines are accessed from the
    Ada.Command_Line.Environment package.  */
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -51,8 +53,6 @@
 #include "system.h"
 #endif
 
-#include "adaint.h"
-
 #ifdef __cplusplus
 extern "C" {
 #endif
diff --git gcc-9-20190324/gcc/ada/cstreams.c gcc-9-20190324/gcc/ada/cstreams.c
index e37070d..f89cbeb 100644
--- gcc-9-20190324/gcc/ada/cstreams.c
+++ gcc-9-20190324/gcc/ada/cstreams.c
@@ -52,6 +52,8 @@
 #include "vxWorks.h"
 #endif
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -61,8 +63,6 @@
 #include "system.h"
 #endif
 
-#include "adaint.h"
-
 #ifdef __cplusplus
 extern "C" {
 #endif
diff --git gcc-9-20190324/gcc/ada/exit.c gcc-9-20190324/gcc/ada/exit.c
index 3ac3596..3fe008e 100644
--- gcc-9-20190324/gcc/ada/exit.c
+++ gcc-9-20190324/gcc/ada/exit.c
@@ -33,6 +33,8 @@
 #include "vxWorks.h"
 #endif
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -42,8 +44,6 @@
 #include "system.h"
 #endif
 
-#include "adaint.h"
-
 #ifdef __cplusplus
 extern "C" {
 #endif
diff --git gcc-9-20190324/gcc/ada/init.c gcc-9-20190324/gcc/ada/init.c
index 00f32e5..83f86f0 100644
--- gcc-9-20190324/gcc/ada/init.c
+++ gcc-9-20190324/gcc/ada/init.c
@@ -53,6 +53,8 @@
 #undef __linux__
 #endif
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -65,7 +67,6 @@
 #include "system.h"
 #endif
 
-#include "adaint.h"
 #include "raise.h"
 
 #ifdef __cplusplus
diff --git gcc-9-20190324/gcc/ada/mkdir.c gcc-9-20190324/gcc/ada/mkdir.c
index d061476..80095c7 100644
--- gcc-9-20190324/gcc/ada/mkdir.c
+++ gcc-9-20190324/gcc/ada/mkdir.c
@@ -34,6 +34,8 @@
 #include <version.h>
 #endif /* __vxworks */
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -53,8 +55,6 @@
 #endif
 #endif
 
-#include "adaint.h"
-
 /*  This function provides a portable binding to the mkdir function.  */
 
 int
diff --git gcc-9-20190324/gcc/ada/raise.c gcc-9-20190324/gcc/ada/raise.c
index 480a0ea..cceeefb 100644
--- gcc-9-20190324/gcc/ada/raise.c
+++ gcc-9-20190324/gcc/ada/raise.c
@@ -32,6 +32,8 @@
 /* Shared routines to support exception handling.  __gnat_unhandled_terminate
    is shared between all exception handling mechanisms.  */
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -40,7 +42,6 @@
 #include "system.h"
 #endif
 
-#include "adaint.h"
 #include "raise.h"
 
 #ifdef __cplusplus
diff --git gcc-9-20190324/gcc/ada/sysdep.c gcc-9-20190324/gcc/ada/sysdep.c
index 21a29bf..5d10f22 100644
--- gcc-9-20190324/gcc/ada/sysdep.c
+++ gcc-9-20190324/gcc/ada/sysdep.c
@@ -52,6 +52,8 @@
 #undef __linux__
 #endif
 
+#include "adaint.h"
+
 #ifdef IN_RTS
 #define POSIX
 #include "tconfig.h"
@@ -72,8 +74,6 @@
 extern struct tm *localtime_r(const time_t *, struct tm *);
 #endif
 
-#include "adaint.h"
-
 /* Don't use macros versions of this functions on VxWorks since they cause
    imcompatible changes in some VxWorks versions */
 #ifdef __vxworks
diff --git gcc-9-20190324/gcc/ada/terminals.c gcc-9-20190324/gcc/ada/terminals.c
index 23f9dfd..6036949 100644
--- gcc-9-20190324/gcc/ada/terminals.c
+++ gcc-9-20190324/gcc/ada/terminals.c
@@ -1150,7 +1150,8 @@ __gnat_setup_winsize (void *desc, int rows, int columns)
 #if defined (__APPLE__) || defined (BSD)
 #define USE_OPENPTY
 #elif defined (__linux__)
-#define USE_GETPT
+#include <pty.h>
+#define USE_OPENPTY
 #elif defined (__sun__)
 #define USE_CLONE_DEVICE "/dev/ptmx"
 #elif defined (_AIX)
